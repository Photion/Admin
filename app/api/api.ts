import axios from 'axios';

import { Model, IModelProps } from '@/app/models/Model';


export class Client {
  static provider = 'api'

  async create(model: Model<IModelProps>) {
    model.created = true;

    const provider = (this.constructor as typeof Client).provider;
    await axios.post(`/${provider}/create`);
  }

  async update(_model: Model<IModelProps>) {
    const provider = (this.constructor as typeof Client).provider;
    await axios.put(`/${provider}/update`);
  }

  async remove(model: Model<IModelProps>) {
    model.created = false;
    const provider = (this.constructor as typeof Client).provider;
    await axios.delete(`/${provider}/remove`);
  }
}
