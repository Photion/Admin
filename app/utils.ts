

export const capitalize = (s: string) => {
  return s
    .replace(/[\W-_]+/g, '-')
    .split('-')
    .map(split => split[0].toUpperCase() + split.substring(1).toLowerCase())
    .join(' ');
};

export const toOption = (value: string) => ({ value, text: capitalize(value) });
