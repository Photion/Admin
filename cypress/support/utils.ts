

export const getFile = (path, mime): Promise<{ file: File; dataTransfer: DataTransfer }> => {
  return new Promise((resolve) => {
    cy.fixture(path, 'base64')
      .then(content => Cypress.Blob.base64StringToBlob(content, mime))
      .then((blob) => {
        const file = new File([blob], path, { type: mime });
        const dataTransfer = new DataTransfer();

        dataTransfer.items.add(file);

        resolve({ file, dataTransfer });
      });
  });
};
