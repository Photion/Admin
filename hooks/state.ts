import { ref, Ref } from '@vue/composition-api';

import { Concept } from '@/app/models/Concept';
import { Media } from '@/app/models/Media';
import { Project } from '@/app/models/Project';

export const media: Ref<{
  concepts: { [key: string]: Concept },
  media: { [key: string]: Media },
  projects: { [key: string]: Project },
}> = ref({
  concepts: {},
  media: {},
  projects: {},
});

export const site = ref({
  domain: '',
  title: '',
  description: '',
});

export const user = ref({
  name: '',
  email: '',
  bio: '',
});

export const useState = () => ({
  media,
  site,
  user,
});
