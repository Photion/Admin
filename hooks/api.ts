
import { ref, Ref, computed } from '@vue/composition-api';
import { HttpClient } from '~/app/api/HttpClient';

export enum Clients {
  '' = 'HttpClient',
  HttpClient = 'HttpClient',
}

type Client = keyof typeof Clients;

export const clients = {
  '': HttpClient,
  HttpClient,
};

export const clientName: Ref<Client> = ref('');

export const client = computed(() => clients[clientName.value]);
