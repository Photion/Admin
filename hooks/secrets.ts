import { ref } from '@vue/composition-api';
import { handler, creds } from '@/hooks/encryption';


export enum Secrets {
  AWS_ACCESS_KEY_ID = 'AWS_ACCESS_KEY_ID',
  AWS_REGION = 'AWS_REGION',
  AWS_SECRET_ACCESS_KEY = 'AWS_SECRET_ACCESS_KEY',
  GCP_API_KEY = 'GCP_API_KEY',
  GCP_API_SECRET = 'GCP_API_SECRET',
}

export const PHOTION_SESSION_CREDENTIALS = 'PHOTION_SESSION_CREDENTIALS';

export const secrets = {
  values: {
    AWS_ACCESS_KEY_ID: ref(''),
    AWS_REGION: ref('eu-west-1'),
    AWS_SECRET_ACCESS_KEY: ref(''),
    GCP_API_KEY: ref(''),
    GCP_API_SECRET: ref(''),
  },

  entries: () => Object.entries(secrets.values),

  list: () => Object.values(secrets.values),

  serialize: () => secrets.entries().map(([key, obj]) => [key, obj.value]),

  deserialize: (payload: { secrets: string[][] }) => {
    payload.secrets.forEach(([key, value]) => {
      const secret = secrets.values[key as Secrets];

      if (secret) {
        secret.value = value;
        return true;
      }

      return false;
    });
  },

  save: async () => {
    const serialized = secrets.serialize();
    window.sessionStorage.setItem(PHOTION_SESSION_CREDENTIALS, JSON.stringify(serialized));

    await handler.save({ password: creds.value }, { secrets: secrets.serialize() });
  },

  load: async () => {
    const payload = await handler.load({ password: creds.value }) as { secrets: string[][] };
    window.sessionStorage.setItem(PHOTION_SESSION_CREDENTIALS, JSON.stringify(payload.secrets));
    secrets.deserialize(payload);
  },

  init: () => {
    const encoded = window.sessionStorage.getItem(PHOTION_SESSION_CREDENTIALS);

    if (encoded) {
      return secrets.deserialize({ secrets: JSON.parse(encoded) });
    }

    return false;
  },
};
