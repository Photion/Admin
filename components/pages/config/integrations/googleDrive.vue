<template>
  <div>
    <h3>Integrate with Google Drive</h3>
    <div>
      Not supported yet.
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api';
import { secrets } from '@/app/state/secrets';

export default defineComponent({

  setup() {
    return {
      secrets,
    };
  },
});
</script>
