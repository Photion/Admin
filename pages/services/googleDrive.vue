<template>
  <service-form :service="googleDrive" preview>
    <div>
      No fields.
    </div>
  </service-form>
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api';

import { secrets } from '@/app/state/secrets';
import { googleDrive } from '@/app/state/service';
import ServiceForm from '@/components/pages/services/ServiceForm.vue';

export default defineComponent({

  components: {
    ServiceForm,
  },

  setup() {
    return {
      googleDrive,
      secrets,
    };
  },
});
</script>
