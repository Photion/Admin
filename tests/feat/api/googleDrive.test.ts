import { TextDecoder } from 'util';

import crypto from '@trust/webcrypto';
import TextEncoding from 'text-encoding-utf-8';

import * as googleDrive from '~/src/api/googleDrive';
import { secrets } from '~/src/state/secrets';
import { getOfflineAccessToken } from '~/tests/utils/googleDrive';

const createFile = (content = {}) => {
  const blob = new Blob([JSON.stringify(content)], { type: 'application/json' });
  const file = new File([blob], 'file.json');

  return file;
};

describe('feat.api.googleDrive', () => {

  beforeAll(async () => {
    global.crypto = crypto;
    (global.TextEncoder as unknown) = TextEncoding.TextEncoder;
    (global.TextDecoder as unknown) = TextDecoder;

    secrets.googleDrive.accessToken = await getOfflineAccessToken();
  });

  afterEach(async () => {
    await googleDrive.clearGoogleDrive();
    secrets.googleDrive.rootFolderId = '';
    secrets.googleDrive.databaseId = '';
  });

  describe('api calls', () => {
    it('Uploads a file and deletes a file', async () => {
      const metadata = {
        name: 'file.json',
      };

      const createRequest = await googleDrive.createFile(metadata, createFile());

      const headRequest1 = await googleDrive.headFile(createRequest.data.id);
      expect(Boolean(headRequest1)).toEqual(true);

      await googleDrive.deleteFile(createRequest.data.id);

      const headRequest2 = await googleDrive.headFile(createRequest.data.id);
      expect(Boolean(headRequest2)).toEqual(false);
    });
  });

  describe('GoogleDriveClient', () => {

    describe('Sustains two concurrent GoogleDriveClients', () => {

      it('Create doesn\'t overwrite', async () => {
        const client1 = new googleDrive.GoogleDriveClient();
        const client2 = new googleDrive.GoogleDriveClient();

        await client1.create('concepts', { uuid: 'concept-1', data: 'abc' });
        await client2.create('concepts', { uuid: 'concept-2', data: 'def' });

        expect(client2.db).toMatchSnapshot();
      }, 20 * 1000);
    });

    describe('Retrieves a saved entity', () => {

      it('Creates with one client, retrieves with another', async () => {
        const client1 = new googleDrive.GoogleDriveClient();
        await client1.create('concepts', { uuid: 'concept-1', data: 'abc' });

        const client2 = new googleDrive.GoogleDriveClient();
        const concept = await client2.retrieve('concepts', 'concept-1');

        expect(concept).toMatchSnapshot();

      }, 20 * 1000);
    });
  });

});
