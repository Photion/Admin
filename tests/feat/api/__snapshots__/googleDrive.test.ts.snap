// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feat.api.googleDrive GoogleDriveClient Retrieves a saved entity Creates with one client, retrieves with another 1`] = `
Object {
  "description": "a",
  "name": "a",
  "portfolio": false,
  "uuid": "folder-1",
}
`;

exports[`feat.api.googleDrive GoogleDriveClient Sustains two concurrent GoogleDriveClients Create doesn't overwrite 1`] = `
Object {
  "folders": Object {},
  "media": Object {},
  "projects": Object {
    "folder-1": Object {
      "description": "a",
      "name": "a",
      "portfolio": false,
      "uuid": "folder-1",
    },
    "folder-2": Object {
      "description": "b",
      "name": "b",
      "portfolio": false,
      "uuid": "folder-2",
    },
  },
}
`;
